src/cocotb/types/__init__.py

```diff
@@ -46,3 +64,117 @@ def concat(a: typing.Any, b: typing.Any) -> typing.Any:
     raise TypeError(
         f"cannot concatenate {type_a.__qualname__!r} with {type_b.__qualname__!r}"
     )
+
+
+from .range import Range  # noqa: F401
+
+
+class ArrayLike(Concatable[T], Protocol, Generic[T]):
+    @property
+    def left(self) -> int:
+        """Leftmost index of the array."""
+        return self.range.left
+
+    @property
+    def direction(self) -> str:
```

Comment:
Maybe this could return something like a Py3.11 `StrEnum`?

Reply:
Something like this?

```python
class Direction(str):
    def __new__(cls: Type[Self], value: str) -> Self:
        value = value.lower()
        if value not in ("to", "downto"):
            raise ValueError(f"{value!r} not one of the possible enumerations ('to', 'downto')")
        return super().__new__(cls, value)

TO = Direction("to")
DOWNTO = Direction("downto")
```

String interning means we don't have to work about singletons and whatnot.

Reply:
Not sure if it's worth the effort, but maybe it helps catching some problems earlier.
About identity... really no idea :)

Reply:
It might be, but that's definitely a follow-up PR.